File: src/euler_011.jl
data = open("dat/euler_011.txt") do file
    readlines(file)
end
data = map(split, data)

M = zeros(Int64, 20, 20)
for i = 1:20
    for j = 1:20
        M[i, j] = parse(Int64, data[i][j])
    end
end

idxs = Base.Iterators.product(1:20, 1:20) |> collect |> vec
out_of_bounds = x -> all(map(y -> 0 < y[1] <= 20 && 0 < y[2] <= 20, x))

path_a = (x, y) -> (map(z -> CartesianIndex(x, y + z), 0:3))
path_b = (x, y) -> (map(z -> CartesianIndex(x + z, y), 0:3))
path_c = (x, y) -> (map(z -> CartesianIndex(x + z, y + z), 0:3))
path_d = (x, y) -> (map(z -> CartesianIndex(x - z, y + z), 0:3))

function answer(matrix)
    m = 0
    for path in (path_a, path_b, path_c, path_d)
        idx = filter(out_of_bounds, map(x -> path(x...), idxs))
        m = max(m, maximum(map(x -> reduce(*, matrix[x]), idx)))
    end
    return m
end

answer(M) |> print

File: src/euler_011.hs
import Data.List

processInput :: String -> [[Int]]
processInput = map (map read . words) . lines

directions = [(0,1),(1,0),(1,1),(1,-1)]
move (x,y) (dx,dy) = (x+dx,y+dy)

answer matrix =
  maximum [ product (takeLine x y dir) | x <- [0..17], y <- [0..17], dir <- directions ]
  where
    takeLine x y dir =
      [ matrix !! x' !! y' | (x',y') <- take 4 $ iterate (move dir) (x,y), inRange x' y' ]

inRange x y = x >= 0 && y >= 0 && x <= 19 && y <= 19

main = do
  contents <- processInput <$> readFile "dat/euler_011.txt"
  putStrLn $ show $ answer contents
